contracts/integrations/custom/CustomIntegrationBalancerv2.sol

```diff
@@ -0,0 +1,416 @@
+     * hparam  _tokensIn                 Addresses of tokens to send to spender to enter
+     * hparam  _maxAmountsIn             Amounts of tokens to send to spender
+     *
+     * @return address                   Target contract address
+     * @return uint256                   Call value
+     * @return bytes                     Trade calldata
+     */
+    function _getEnterCalldata(
+        address _strategy,
+        bytes calldata _data,
+        uint256 _resultTokensOut,
+        address[] calldata _tokensIn,
+        uint256[] calldata _maxAmountsIn
+    )
+        internal
+        view
+        override
+        returns (
+            address,
+            uint256,
+            bytes memory
+        )
+    {
+        bytes32 poolId = BasePool(BytesLib.decodeOpDataAddress(_data)).getPoolId();
+        address strategy = _strategy;
+        uint256 resultTokensOut = _resultTokensOut;
+
+        (IERC20[] memory poolTokens, , ) = IVault(vaultAddress).getPoolTokens(poolId);
+        require(_tokensIn.length == poolTokens.length, 'Must supply same number of tokens as are already in the pool!');
+
+        IVault.JoinPoolRequest memory joinRequest = _getJoinRequest(
+            _tokensIn,
+            poolTokens,
+            _maxAmountsIn,
+            resultTokensOut
+        );
+        bytes memory methodData = abi.encodeWithSelector(
+            IVault.joinPool.selector,
+            poolId,
+            strategy,
+            strategy,
+            joinRequest
+        );
+
+        return (vaultAddress, 0, methodData);
+    }
+
+    /**
+     * Return exit custom calldata
+     *
+     * @param  _strategy                 Address of the strategy
+     * @param  _data                     OpData e.g. Address of the pool
+     * @param  _resultTokensIn           Amount of result tokens to send
+     * @param  _tokensOut                Addresses of tokens to receive
+     * @param  _minAmountsOut            Amounts of input tokens to receive
+     *
+     * @return address                   Target contract address
+     * @return uint256                   Call value
+     * @return bytes                     Trade calldata
+     */
+    function _getExitCalldata(
+        address _strategy,
+        bytes calldata _data,
+        uint256 _resultTokensIn,
+        address[] calldata _tokensOut,
+        uint256[] calldata _minAmountsOut
+    )
+        internal
+        view
+        override
+        returns (
+            address,
+            uint256,
+            bytes memory
+        )
+    {
+        bytes32 poolId = BasePool(BytesLib.decodeOpDataAddress(_data)).getPoolId();
+        address strategy = _strategy;
+        uint256 resultTokensIn = _resultTokensIn;
+
+        (IERC20[] memory tokens, , ) = IVault(vaultAddress).getPoolTokens(poolId);
+        require(_tokensOut.length == tokens.length, 'Must supply same number of tokens as are already in the pool!');
```

Comment:
is this require needed?

Reply:
This is the same require as in line 141. None of these are necessary, if we assume that the strategy will always correctly get the tokens and their min/max amounts from the respective functions of our integration. Do you think we should remove both the require in line 141 and here?